app/models/email.rb

```diff
@@ -0,0 +1,30 @@
+# frozen_string_literal: true
+
+class Email < ApplicationRecord
+  class Association < ApplicationRecord
+    belongs_to :email
+    belongs_to :object, polymorphic: true
+  end
+
+  has_many :associations, dependent: :destroy
+
+  UNDEFINED = Object.new
+  private_constant :UNDEFINED
+
+  scope(:associated_with, lambda do |object, as: UNDEFINED|
+    association_scope = Association.where(object: object)
+    association_scope = association_scope.where(name: as) unless as == UNDEFINED
+
+    where(id: association_scope.select(:email_id))
+  end)
```

Comment:
1. Why do we need an undefined object? Can't we use `as: nil` and have a guard `if as`? 
2. That scope brackets look pretty awkward, especially closing one after an `end`. Can we get rid of that?

Reply:
1. `as` is optional - `associated_with(object, as: ni)l` means "give me emails that has been associated without any name` while `associated_with(object)` means "give me all emails, regardless of association name".
2. Yeah, I was very unhappy about this as well, but this is how RuboCop want to see it becasue:
  - `lambda` must be used over `->` for multiline blocks (don't really agree with it!) 
  - `scope :associated_with, lambda do |v|` passes block with `scope` method rather than `lambda` (since `lambda` is a method and not a ruby code token) - hence the need for brackets
Happy to disable rubocop lambda check here!

Reply:
> `associated_with(object)` means "give me all emails, regardless of association name"

You mean the same as `Email.all`?

Reply:
No, it will return all emails that are associated with an object...

Reply:
Right, ok. So we would need the name when the same object has been associated as two different roles for some emails. 

And one edge case is we want emails where the object is a specific role - then we pass a non-default name.

Another edge case is we want all emails for the object - and at the moment we pass in nothing, and the default value does the trick.

What I struggle to understand is: why would we want the emails where the association name is specifically `nil`. Surely each association we create has a meaningful name - or at least we can give it some name that is better than nothing. 

So I guess it boils down to "should we make `name` of the association not nullable, and when we receive a nil name in this scope just return everything for this object"?